src/flint/FlintCTypes.jl

```diff
@@ -0,0 +1,1108 @@
+
+@flintstruct struct fq_poly_struct
+  coeffs::Ptr{fq_struct}
+  alloc::slong
+  length::slong
+end
+
+const fq_poly_t = Tuple{fq_poly_struct}
+
+@flintstruct struct fq_poly_factor_struct
+  poly::Ptr{fq_poly_struct}
+  exp::Ptr{slong}
+  num::slong
+  alloc::slong
+end
+
+const fq_poly_factor_t = Tuple{fq_poly_factor_struct}
+
+# end fq_types.h
+###############################################################################
+
+
+###############################################################################
+# begin gr_types.h
+
+const GR_SUCCESS = 0
+
+const GR_DOMAIN = 1
+
+const GR_UNABLE = 2
+
+const GR_TEST_FAIL = 4
+
+@enum truth_t begin
+  T_TRUE
+  T_FALSE
+  T_UNKNOWN
+end
+
+@flintstruct struct gr_stream_struct
+  fp::Ptr{FLINT_FILE}
+  s::Ptr{Cchar}
+  len::slong
+  alloc::slong
+end
+
+const gr_stream_t = Tuple{gr_stream_struct}
+
+const gr_funcptr = Ptr{Cvoid}
+
+const GR_CTX_STRUCT_DATA_BYTES = 6 * sizeof(ulong)
+
+@flintstruct struct gr_ctx_struct
+  data::NTuple{GR_CTX_STRUCT_DATA_BYTES, Cchar}
+  which_ring::ulong
+  sizeof_elem::slong
+  methods::Ptr{gr_funcptr}
+  size_limit::ulong
+end
+
+const gr_ctx_t = Tuple{gr_ctx_struct}
+
+const gr_ptr = Ptr{Cvoid}
+
+const gr_srcptr = Ptr{Cvoid}
+
+const gr_ctx_ptr = Ptr{Cvoid}
+
+@flintstruct struct gr_vec_struct
+  entries::gr_ptr
+  alloc::slong
+  length::slong
+end
+
+const gr_vec_t = Tuple{gr_vec_struct}
+
+@flintstruct struct gr_mat_struct
+  entries::gr_ptr
+  r::slong
+  c::slong
+  rows::Ptr{gr_ptr}
+end
+
+const gr_mat_t = Tuple{gr_mat_struct}
+
+@flintstruct struct gr_poly_struct
+  coeffs::gr_ptr
+  alloc::slong
+  length::slong
+end
+
+const gr_poly_t = Tuple{gr_poly_struct}
+
+# end gr_types.h
+###############################################################################
+
+
+###############################################################################
+# begin fq_default.h
+
+const FQ_DEFAULT_FQ_ZECH = 1
+
+const FQ_DEFAULT_FQ_NMOD = 2
+
+const FQ_DEFAULT_FQ = 3
+
+const FQ_DEFAULT_NMOD = 4
+
+const FQ_DEFAULT_FMPZ_MOD = 5
+
+struct fq_default_struct
+  uniondata::NTuple{maximum(sizeof, (
+    fq_t,
+    fq_nmod_t,
+    fq_zech_t,
+    ulong,
+    fmpz_t,
+  )), UInt8}
+end
+const union_fq_default_struct = fq_default_struct
+
+const fq_default_t = Tuple{fq_default_struct}
+
+const fq_default_ctx_struct = gr_ctx_struct
+
+const fq_default_ctx_t = Tuple{fq_default_ctx_struct}
+
+@flintstruct struct _gr_fmpz_mod_ctx_struct
+  ctx::Ptr{fmpz_mod_ctx_struct}
+  is_prime::truth_t
+  a::fmpz
+end
+
+@flintstruct struct _gr_nmod_ctx_struct
+  nmod::nmod_t
+  a::ulong
+end
+
+# end fq_default.h
+###############################################################################
+
+
+###############################################################################
+# begin fq_default_mat.h
+
+struct fq_default_mat_struct
+  uniondata::NTuple{maximum(sizeof, (
+    fq_mat_t,
+    fq_nmod_mat_t,
+    fq_zech_mat_t,
+    nmod_mat_t,
+    fmpz_mod_mat_t,
+  )), UInt8}
```

Comment:
Just wanted to mention that I'd been thinking about doing something akin to this (separate from this PR) to allow more optimized code:
```suggestion
  entries::Ptr{Nothing}
  r::slong
  c::slong
  rows::Ptr{Nothing}
  uniondata::NTuple{maximum(sizeof, (
    fq_mat_t,
    fq_nmod_mat_t,
    fq_zech_mat_t,
    nmod_mat_t,
    fmpz_mod_mat_t,
  ) - 32), UInt8}
```
(where 32 of course is `2*sizeof(Ptr{Nothing}) + 2*sizeof(slong)`).

So that one can then write fast accessors to these fields. Similar for `fq_default_poly`.

Not sure how to fit this with this PR... but one option would be to automate this as part of Union handling: if forming the C union of C struct types $T_1, \dots, T_k$ then we should in principle be able to figure out if they have common entries at the start, and automatically "extract" them (with slight cheating to "merge" `Ptr` types).